File: Extensions/engine/chunks.py
from posixpath import abspath
import sys

sys.path.append("..\Extensions")
from geometry import Cube

from os import makedirs
from os.path import abspath, exists, join, isfile
from pyrr import Vector3
from numpy import float16, uint8
from numpy import zeros, array, append, save, load



BlockTypes = {
    0: [],
    1: [3, 3, 3, 3, 2, 0],
}

WORLD_PATH = abspath('Worlds/1st_world/')
if not exists(WORLD_PATH):
    makedirs(WORLD_PATH)



class Chunk:
    CHUNK_SIZE = Vector3([16, 128, 16], dtype=uint8)
    CHUNK_SIZE_yz = CHUNK_SIZE.y * CHUNK_SIZE.z
    CUBE = Cube()

    def __init__(self, atlas, chunk_position: tuple[3] or list[3] = (0, 0, 0)):
        self.position = array(chunk_position)

        self.atlas = atlas
        self.vertices = zeros(0, float16)
        self.chunk_data = zeros(0)


    def generate_chunk(self, height: int or float):
        height = min(height, Chunk.CHUNK_SIZE.y)
        chunk_data = zeros(0)

        for x in range(Chunk.CHUNK_SIZE.x):
            for y in range(Chunk.CHUNK_SIZE.y):
                for z in range(Chunk.CHUNK_SIZE.z):
                    if y < height:
                        # append block - block type (default: 1)
                        data = 1
                    else:
                        # append air - 0
                        data = 0

                    chunk_data = append(chunk_data, data)

        self.chunk_data = chunk_data

    @property
    def grid_position(self):
        return self.position // Chunk.CHUNK_SIZE

    @staticmethod
    def data_index_to_position(index):
        x = index // Chunk.CHUNK_SIZE_yz
        y = index // Chunk.CHUNK_SIZE.z
        z = index % Chunk.CHUNK_SIZE_yz

        return x, y, z


    def generate_mesh(self):
        # 4 vertices per each face
        # 3 * 2 triangle indices per each face

        for x in range(Chunk.CHUNK_SIZE.x):
            for y in range(Chunk.CHUNK_SIZE.y):
                for z in range(Chunk.CHUNK_SIZE.z):
                    # block variables
                    block_position = Vector3([x, y, z])
                    block_type = self.chunk_data[block_position.x * Chunk.CHUNK_SIZE_yz + block_position.y * Chunk.CHUNK_SIZE.z + block_position.z]
                    if block_type:
                        block_type_array = BlockTypes[block_type]

                    if block_type == 0:
                        continue

                    # check for each face if is exposed (next to 0 or out of limit)
                    for face_index in range(0, 6):
                        face_normal = Cube.normals[face_index]
                        check_position = block_position + face_normal

                        if 0 <= check_position.x < Chunk.CHUNK_SIZE.x and 0 <= check_position.y < Chunk.CHUNK_SIZE.y and 0 <= check_position.z < Chunk.CHUNK_SIZE.z:
                            block_data = self.chunk_data[check_position.x * Chunk.CHUNK_SIZE_yz + check_position.y * Chunk.CHUNK_SIZE.z + check_position.z]

                            if block_data == 0:
                                self.generate_face(block_position, face_index, block_type_array[face_index])
                        else:
                            self.generate_face(block_position, face_index,block_type_array[face_index])


    def generate_face(self, block_position, face_index, texture:int):
        # block_pos-3 places(can be converted to 1 number), face_index-1place, atlas_offset-2 places
        data = [*(block_position), face_index, *self.atlas.get_position(texture)]
        data = array(data, dtype=float16)

        self.vertices = append(self.vertices, data)



class ChunkManager:
    FAILED = -1
    SUCCESS = 0

    @staticmethod
    def save(position: tuple[3] or list[3], data):
        with open(join(WORLD_PATH, f'{position[0]}_{position[2]}.npy'), 'w') as f:
            save(f, data)
            return ChunkManager.SUCCESS


    @staticmethod
    def load(position: tuple[3] or list[3], atlas):
        file_path = join(WORLD_PATH, f'{position[0]}_{position[2]}.npy')
        chunk = Chunk(atlas, position * Chunk.CHUNK_SIZE)

        if not exists(file_path):
            # create chunk
            chunk.generate_chunk(10)
        else:
            with open(file_path, 'r') as f:
                data = load(f)
                chunk.chunk_data = data
        
        chunk.generate_mesh()
        return chunk

File: Extensions/engine/c++/chunks/chunks.cpp
#include "chunks.h"

Vector3<int> chunkSize(16, 256, 16);
Vector3<int> chunkSizeVec(256 * 16, 16, 1);
Vector3<int> *chunkSizePtr = &chunkSize;
Vector3<int> *chunkSizeVecPtr = &chunkSizeVec;

int blockTypes[2][6] = {
    {1, 1, 1, 1, 1, 1},
    {3, 3, 3, 3, 2, 0}};

int normals[6][3] = {
    0, 0, 1,
    0, 0, -1,
    1, 0, 0,
    -1, 0, 0,
    0, -1, 0,
    0, 1, 0};

void setChunkSize(Vector3<int> &newSize)
{
    *chunkSizePtr = newSize;
    chunkSizeVecPtr = new Vector3<int>(newSize.y * newSize.z, newSize.z, 1);
}

std::vector<int> generateChunkData()
{
    int height = 10;
    std::vector<int> chunkData;
    chunkData.reserve(chunkSize.x * chunkSize.y * chunkSize.z);

    for (int x = 0; x < chunkSize.x; x++)
    {
        for (int y = 0; y < chunkSize.y; y++)
        {
            for (int z = 0; z < chunkSize.z; z++)
            {
                if (y <= height)
                    chunkData.emplace_back(1);
                else
                    chunkData.emplace_back(0);
            }
        }
    }

    return chunkData;
}

std::vector<int *> generateMeshData(int (&chunkData)[])
{
    // 4 vertices per each face
    // 3 * 2 triangle indices per each face
    std::vector<int *> vertices;

    for (int x = 0; x < chunkSize.x; x++)
    {
        for (int y = 0; y < chunkSize.y; y++)
        {
            for (int z = 0; z < chunkSize.z; z++)
            {
                // block variables
                const Vector3<int> blockPosition(x, y, z);
                const int &index = blockPosition | chunkSizeVec;
                int blockType = chunkData[index];
                int(&blockTypeFaces)[6] = blockTypes[blockType];

                if (blockType == 0)
                    continue;

                for (int faceIndex = 0; faceIndex < 6; faceIndex++)
                {
                    Vector3<int> &faceNormal = *new Vector3<int>(normals[faceIndex]);
                    Vector3<int> &checkPosition = blockPosition + faceNormal;

                    if (0 <= checkPosition.x < chunkSize.x && 0 <= checkPosition.y < chunkSize.y && 0 <= checkPosition.z < chunkSize.z)
                    {
                        const int &checkIndex = checkPosition | chunkSizeVec;
                        int *checkBlockType = &chunkData[checkIndex];

                        if (!checkBlockType)
                            vertices.push_back({GenerateFace(blockPosition, faceIndex, blockTypeFaces[faceIndex])});
                    }
                    else
                        vertices.push_back({GenerateFace(blockPosition, faceIndex, blockTypeFaces[faceIndex])});
                }
            }
        }
    }
    return vertices;
}

int *GenerateFace(Vector3<int> blockPosition, int faceIndex, int textureIndex)
{
    int *data[] = {&blockPosition.x, &blockPosition.y, &blockPosition.z, &faceIndex, &textureIndex};

    return *data;
}

std::vector<int *> loadChunk(Vector3<int> &position)
{
    char path_buffer[50];
    sprintf(path_buffer, "chunks/%d_%d.txt", position.x, position.z);

    std::ifstream file(path_buffer);
    std::vector<int> chunkData;

    if (!file.is_open())
    {
        // no file
        std::cout << "There's no file" << std::endl;

        chunkData = generateChunkData();
    }
    else
    {
        // read file
        std::cout << "Read file" << std::endl;
        std::string fileText;
        chunkData.reserve(chunkSize.x * chunkSize.y * chunkSize.z);
        const int dataSize = chunkSize.x * chunkSize.z;
        //std::string data[data_size];

        while (std::getline(file, fileText, ' '))
        {
            // Output the text from the file
            std::cout << fileText << std::endl;
            chunkData.emplace_back(fileText);
        }

        file.close();
    }

    std::vector<int *> meshData = generateMeshData(chunkData);

    return meshData;
}

int saveChunk(Vector3<int> &position, std::vector<int> &chunkData)
{
    std::ofstream file;

    char pathBuffer[50];
    sprintf(pathBuffer, "chunks/%d_%d.txt", position.x, position.z);

    std::ofstream file(pathBuffer);

    for (int blockData : chunkData)
    {
        file << blockData << ' ';
    }

    file.close();
    return 0;
}
